src/tcp_orientation_calibration.cpp

```diff
@@ -0,0 +1,182 @@
+    {
+      ROS_WARN("Check to make sure that a robot state publisher or other node is publishing"
+               " tf frames for your robot. Also check that your base/tool frames names are"
+               " correct and not missing a prefix, for example.");
+    }
+    else if (!base_found)
+    {
+      ROS_WARN("Check to make sure that base frame '%s' actually exists.", base_frame.c_str());
+    }
+    else if (!tool_found)
+    {
+      ROS_WARN("Check to make sure that tool0 frame '%s' actually exists.", tool0_frame.c_str());
+    }
+    return 1;
+  }
+  // Create storage for user observations
+  tool_point_calibration::Affine3dVector observations;
+  observations.reserve(num_samples);
+  std::string line;
+  int count = 0;
+  // While ros is ok and there are more counts to be done...
+  while (ros::ok() && count < num_samples)
+  {
+    ROS_INFO("Pose %d: Jog robot to a new location touching the shared position and"
+             " press enter.", count);
+    std::getline(std::cin, line); // Blocks program until enter is pressed
+    tf::StampedTransform transform;
+    try
+    {
+      listener.lookupTransform(base_frame, tool0_frame,
+                               ros::Time(0), transform);
+      Eigen::Affine3d eigen_pose;
+      tf::poseTFToEigen(transform, eigen_pose);
+      observations.push_back(eigen_pose);
+      ROS_INFO_STREAM("Pose " << count << ": captured transform:\n" << eigen_pose.matrix());
+      count++;
+    }
+    catch (const tf::TransformException& ex)
+    {
+      ROS_ERROR("%s",ex.what());
+      ros::Duration(1.0).sleep();
+      continue;
+    }
+  }
+  ROS_INFO("Calibration captured %d tool poses (out of %d requested). Computing calibration...",
+           count, num_samples);
+  tool_point_calibration::TcpCalibrationResult result =
+      tool_point_calibration::calibrateTcp(observations);
+  ROS_INFO_STREAM("Calibrated tcp (meters in xyz): [" << result.tcp_offset.transpose() << "] from " << tool0_frame);
+  ROS_INFO_STREAM("Touch point (meters in xyz): [" << result.touch_point.transpose() << "] in frame " << base_frame);
+  ROS_INFO_STREAM("Average residual: " << result.average_residual);
+  ROS_INFO_STREAM("Converged: " << result.converged);
+  // store the translation vector for the next step:
+  Vector3d tcp_transl;
+  tcp_transl = result.tcp_offset.transpose();
+  if (count < 4)
+  {
+    ROS_WARN("Computing a tool calibration w/ fewer than 4 points may produce an answer with good"
+             " convergence and residual error, but without a meaningful result. You are encouraged"
+             " to try with more points");
+  }
+  ROS_INFO("=======\n step 2: TCP Orientation calibration using three point method ========== ");
+  std::vector<Affine3d> A0n; // to store the transformation matrix for the points
+  A0n.resize( 3);
+
+  for (int i=0; i<3; i++)
+  {
+    if(i ==0)
+        ROS_INFO("First point: keep the TCP at same point where the translation calibration was done, press enter");
```

Comment:
Is this step really needed? Perhaps incorporate the warning in the previous step and just move to the "Move along axis .. " step immediately.

Reply:
this step is similar to the next two steps  i=1,2. I thought it is better to combine 3 of them inside one loop instead of writing the body of the loop [try, catch statement] once again in the previous step.

Reply:
I meant more someting like this, where the collection step at i == 0 is taken out of the loop. As it has already been captured it doesn't need the overhead of the try-catch and conversion to an eigen vector:
```
A0n[0] = observations[num_samples-1];
for (int i=1; i<3; i++) 
{
 ...
}
```

Reply:
done!